File: 20. Valid Parentheses/validparentheses.java
import java.util.*;

class validparentheses {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<Character>();
        HashMap<Character, Character> hm = new HashMap<>();
        hm.put('(', ')');
        hm.put('[', ']');
        hm.put('{', '}');

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (hm.containsKey(c)) {
                st.push(c);
            } else {
                if (st.isEmpty() || hm.get(st.peek()) != c) {
                    return false;
                }
                st.pop();
            }
        }
        return st.isEmpty();
    }
}

// validparentheses solution = new validparentheses();
// System.out.println(solution.isValid("[]()"));

File: 20. Valid Parentheses/validparentheses.py
class Solution(object):
    def isValid(self, s):
        d = {'(': ')',
            '[':']',
            '{': '}'
        }
        st = list()
        for char in s:
            if not st or char in d: 
                st.append(char)
                continue
            if d[st[-1]] == char:
                st.pop()
            else:
                st.append(char)
        return len(st) == 0

# s = Solution()
# print(s.isValid("()[]"))